### PR TITLE
Encapsulate HeapMonitor logic

### PR DESCRIPTION
**Pull request by dfquaresma**
This PR ports HeapMonitor logic from [gci-go heap](https://github.com/gcinterceptor/gci-go/blob/master/gccontrol/runtime.go#L50) to gci-java.

### CODE REVIEW

File: gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package com.gcinterceptor.gci;
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.MemoryPoolMXBean;
+
+public class HeapMonitor {
```
Comment (danielfireman) [inline gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java]: No need for this class to be public.

File: gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	private MemoryPoolMXBean youngPool;
+	private GarbageCollector collector;
+	private long lastAlloc; 
+	private long lastUsed;
+
+	public HeapMonitor() {
+		for (final MemoryPoolMXBean pool : ManagementFactory.getMemoryPoolMXBeans()) {
+			// TODO(danielfireman): Generalize this to other JVM versions.
```
Comment (danielfireman) [inline gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java]: I know why this is here. Feel free to remove this TODO.

File: gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+			if (pool.getName().contains("Eden")) {
+				youngPool = pool;
+				break;
+			}
+		}
+		lastAlloc = getUsage();
+		collector = () -> System.gc();
+	}
+
+	Long getUsage() {
```
Comment (danielfireman) [inline gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java]: this must be private.

Comment (danielfireman) [inline gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java]: The return type should be long.

File: gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+		return  this.youngPool.getUsage().getUsed();
+	}
+
+	Long AllocSinceLastGC() {
```
Comment (danielfireman) [inline gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java]: return type should be long

Comment (danielfireman) [inline gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java]: please rename to getHeapUsageSinceLastGC .. In java, method names start with uncapitalized characters.

File: gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+		lastUsed = getUsage() - lastAlloc;
+		return lastUsed;
+	}
+	
+	Long collect() {
```
Comment (danielfireman) [inline gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java]: return type shold be long.

File: gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
+package com.gcinterceptor.gci;
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.MemoryPoolMXBean;
+
+public class HeapMonitor {
+	private MemoryPoolMXBean youngPool;
+	private GarbageCollector collector;
```
Comment (danielfireman) [inline gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java]: Maybe you forgot to add this class to the PR?

File: gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+	private long lastAlloc; 
+	private long lastUsed;
+
+	public HeapMonitor() {
```
Comment (danielfireman) [inline gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java]: no need of this method to be public.

File: gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+		for (final MemoryPoolMXBean pool : ManagementFactory.getMemoryPoolMXBeans()) {
+			// TODO(danielfireman): Generalize this to other JVM versions.
+			if (pool.getName().contains("Eden")) {
+				youngPool = pool;
+				break;
+			}
+		}
+		lastAlloc = getUsage();
+		collector = () -> System.gc();
```
Comment (danielfireman) [inline gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java]: NIce!

File: gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java (hunk 2 above)
Comment (dfquaresma) [reply gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java]: Done

File: gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java (hunk 1 above)
Comment (dfquaresma) [reply gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java]: Done

File: gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java (hunk 3 above)
Comment (dfquaresma) [reply gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java]: Done

File: gci/src/main/java/com/gcinterceptor/gci/GarbageCollector.java [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package com.gcinterceptor.gci;
+
+/**
+ * Abstract garbage the runtime's garbage collector. Meant to be used by tests.
+ *
+ * @author danielfireman
+ */
+@FunctionalInterface
+public interface GarbageCollector {
```
Comment (danielfireman) [inline gci/src/main/java/com/gcinterceptor/gci/GarbageCollector.java]: No need for a file for this interface. It is totally fine for this to be part of the HeapMonitor. Also, please make it private.

File: gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
+package com.gcinterceptor.gci;
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.MemoryPoolMXBean;
+
+class HeapMonitor {
```
Comment (danielfireman) [inline gci/src/main/java/com/gcinterceptor/gci/HeapMonitor.java]: This class is not only monitoring anymore, it does collections. So, let's call it Heap (remember you need to rename the file).

**Pull request merged**